File: functions/optimizationNewton.m
function [rho_opt, E0_max] = optimizationNewton(Q, pi_matrix, g_matrix, R, tol, rho_star)
%   Optimize the E0 function using Newton's method.
%
%   [rho_opt, E0_max] = optimizationNewton(Q, pi_matrix, g_matrix, R, tol, rho_star)
%
%   This function applies Newton's method to find the optimal value of the parameter
%   rho that maximizes the function E0, whose value and derivatives are computed by the
%   function 'derivativesE0'. The iteration updates rho using the gradient and curvature
%   of E0 until the change in rho is smaller than the specified tolerance or a maximum
%   number of iterations is reached.
%
%   Inputs:
%       Q         - Matrix or data structure required for computing E0.
%       pi_matrix - Matrix of probability values or related parameters for the E0 function.
%       g_matrix  - Matrix containing gain values or similar parameters used in E0's computation.
%       R         - Scalar parameter representing a rate or threshold used in the optimization.
%       tol       - Tolerance for the convergence criterion based on the change in rho.
%       rho_star  - Initial guess for the value of rho.
%
%   Outputs:
%       rho_opt   - The optimized value of rho at which E0 is maximized.
%       E0_max    - The value of the function E0 evaluated at rho_opt.
%

    iter = uint32(0);
    log2_const = log(2);
    rho = rho_star;
    max_iter = 100;
    tol_curvature = 1e-10;
    
    while (iter < max_iter)
        % Compute all derivatives in one pass to avoid redundant calculations
        [F0, E0, dF0, ~, d2F0, ~] = derivativesE0(Q, pi_matrix, g_matrix, rho);

        % Optimize divisions by precomputing repeated values
        f_inv = 1/F0;
        fp_f = dF0 * f_inv;

        % Compute gradient and curvature with minimized operations
        gradient = (-fp_f / log2_const) - R;
        curvature = -((d2F0 * f_inv) - (fp_f * fp_f)) / log2_const;
        
        % Combine convergence checks to reduce branching
        if abs(gradient/curvature) < tol || abs(curvature) < tol_curvature
            break;
        end
        
        disp(rho);

        % Compute new rho value and clamp to [0, 1]
        rho_new = rho - gradient / curvature;
    
        % Check convergence based on rho change
        if abs(rho_new - rho) < tol
            break;
        end
    
        rho = rho_new;
        iter = iter + 1;
    end

    % Final output:
    rho_opt = rho;
    % E0(rho_opt) (the function value at final rho):
    E0_max = E0;

    fprintf('\nNewton algorithm finished after %u iterations.\n', iter);
end

File: functions/python/optimizationNewton.py
import numpy as np
from functions.python.derivativesE0 import derivativesE0

def optimizationNewton(Q, pi_matrix, g_matrix, R, tol, rho_star):
    """
    Optimize the E0 function using Newton's method.

    Parameters:
        Q         : 1D array or similar data structure required for computing E0.
        pi_matrix : 2D array of probability-related values for E0.
        g_matrix  : 2D array containing gain values or similar parameters for E0.
        R         : Scalar parameter representing a rate or threshold used in the optimization.
        tol       : Tolerance for the convergence criterion based on the change in rho.
        rho_star  : Initial guess for the value of rho.

    Returns:
        rho_opt   : The optimized value of rho at which E0 is maximized.
        E0_max    : The value of the function E0 evaluated at rho_opt.
    """
    iter = 1
    log2_const = np.log(2)
    rho = rho_star
    max_iter = 100
    tol_curvature = 1e-10

    while iter < max_iter:
        # Compute F0, dF0, d2F0 from derivativesE0 function.
        F0, dF0, d2F0 = derivativesE0(Q, pi_matrix, g_matrix, rho)
        E0 = -np.log2((1/np.pi) * F0)

        # Precompute repeated terms to optimize divisions.
        f_inv = 1 / F0
        fp_f = dF0 * f_inv

        # Compute gradient and curvature.
        gradient = (-fp_f / log2_const) - R
        curvature = -((d2F0 * f_inv) - (fp_f * fp_f)) / log2_const

        # Convergence check based on the Newton update or small curvature.
        if abs(gradient / curvature) < tol or abs(curvature) < tol_curvature:
            break

        # Compute new rho value using Newton's update.
        rho_new = rho - gradient / curvature

        # Check convergence based on the change in rho.
        if abs(rho_new - rho) < tol:
            rho = rho_new
            break

        rho = rho_new
        iter += 1

    rho_opt = rho
    E0_max = E0

    print(f"rho_opt = {rho_opt:.6f}")
    print(f"E0_max = {E0_max:.6f}")
    print(f"iter = {iter}")
    
    return rho_opt, E0_max